### PR TITLE
GODRIVER-1705 Add versioned API

### PR DESCRIPTION
**Pull request by benjirewis**
[GODRIVER-1705](https://jira.mongodb.org/browse/GODRIVER-1705)

Implements versioned API for the Go driver according to [spec](https://docs.google.com/document/d/1Hg4Gl_WA6TE6ZXHpULau2C6SOEtHzi9u2hBJ3NphM9c/edit#). See the [original draft PR](https://github.com/mongodb/mongo-go-driver/pull/551) for older, iterative review/discussion.

Overarching changes:
- Updates unified spec test runner to handle schema  1.1
- Creates new ServerAPIOptions on client options
- Appends versioned API fields to most commands sent to the server
- Updates Evergreen configs to run new tests against specially configured servers

### CODE REVIEW

File: .evergreen/config.yml [hunk 1]
```diff
@@ -1528,6 +1556,37 @@ tasks:
...
+          SSL: "nossl"
+          REQUIRE_API_VERSION: true
+      - func: run-versioned-api-test
+        vars:
+          TOPOLOGY: "server"
+          AUTH: "auth"
+          SSL: "nossl"
+          REQUIRE_API_VERSION: true
+
+  - name: "test-standalone-versioned-api-test-commands"
```
Comment (benjirewis) [inline .evergreen/config.yml:1575]: This new task runs tests in `./mongo`, `./mongo/integration` and `./mongo/integration/unified` against a standalone with `enableTestCommands=1`, `acceptAPIVersion2=1` and `requireAPIVersion=0`. Technically, this standalone config exists _just_ to run the new spec test `test-commands-deprecation-errors`, so running so many tests might be overkill. Let me know if so; I could make a new `Makefile` recipe that just runs the single test...

File: data/versioned-api/transaction-handling.json [hunk 2]
```diff
@@ -0,0 +1,232 @@
+{
+  "description": "Transaction handling",
+  "schemaVersion": "1.1",
+  "runOnRequirements": [
+    {
+      "minServerVersion": "4.7",
+      "topologies": [
+        "replicaset",
```
Comment (benjirewis) [inline data/versioned-api/transaction-handling.json:8]: This new spec test can only run against a replica set or a sharded cluster, but I still need to enable requireAPIVersion on every node to do so.

File: x/mongo/driver/operation.go [hunk 3]
```diff
@@ -764,6 +767,12 @@ func (op Operation) createQueryWireMessage(dst []byte, desc description.Selected
 		return dst, info, err
 	}
 
+	// Add server API information before calling addSession() because it will modify
+	// transaction state.
+	if op.Client == nil || !op.Client.TransactionInProgress() {
```
Comment (kevinAlbs) [inline x/mongo/driver/operation.go:772]: Transactions that have been committed can be repeatedly committed:
https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#clientsession-changes

The state of the session after the first commit changes to `Committed`. 

I think there could be a danger when double committing a transaction:
- On the first commit, `TransactionInProgress() == true`, no server API fields are appended.
- On the second commit, `TransactionInProgress() == false` returns false on the second commit. So this block is entered.

Updating transaction-handling.json to append an extra commit seems to still pass though. So I may be misunderstanding.

File: mongo/options/serverapioptions.go [hunk 4]
```diff
@@ -62,3 +66,15 @@ func (sav ServerAPIVersion) Validate() error {
 	}
 	return fmt.Errorf("api version %q not supported by driver", sav)
 }
+
+// ConvertToDriverAPIOptions converts a options.ServerAPIOptions instance to a driver.ServerAPIOptions.
+func (s *ServerAPIOptions) ConvertToDriverAPIOptions() *driver.ServerAPIOptions {
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/options/serverapioptions.go:71]: Can do! Moved to `client` as `convertToDriverAPIOptions()`.

File: mongo/integration/unified/server_api_options_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import (
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// ServerAPIOptions is a wrapper for *options.ServerAPIOptions. This type implements the bson.Unmarshaler interface
+// to convert BSON documents to a ServerAPIOptions instance.
+type ServerAPIOptions struct {
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/integration/unified/server_api_options_test.go:18]: Yep! Funny Go things :)

File: x/mongo/driver/operation.go (hunk 3 above)
Comment (benjirewis) [reply x/mongo/driver/operation.go:772]: I agree! I'll bring it up in the versioned-api implementation channel before submitting a PR but will link it in this thread if/when I end up making one.

**Review by kevinAlbs — APPROVED** [review]
Phenomenal work, LGTM!

**Review by iwysiu — APPROVED** [review]

File: mongo/database.go [hunk 6]
```diff
@@ -180,6 +180,9 @@ func (db *Database) processRunCommand(ctx context.Context, cmd interface{},
 // If the command document contains a session ID or any transaction-specific fields, the behavior is undefined.
 //
 // The opts parameter can be used to specify options for this operation (see the options.RunCmdOptions documentation).
+//
+// Specifying API versioning options in the command document and declaring an API version on the client is not supported.
```
Comment (divjotarora) [inline mongo/database.go:184]: I'd move this up to the second paragraph as it already discusses a similar case for session/transaction fields.

File: mongo/options/serverapioptions.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+// ServerAPIOptions represents options used to configure the API version sent to the server
+// when running commands.
+//
+// Sending a specified server API version causes the server to behave in a manner compatible with that
+// API version. It also causes the driver to behave in a manner compatible with the driver’s behavior as
+// of the release when the driver first started to support the specified server API version.
+//
+// The user must specify a ServerAPIVersion if including ServerAPIOptions in their client. That version
```
Comment (divjotarora) [inline mongo/options/serverapioptions.go:20]: Should we explicitly enumerate the supported versions in this paragraph?

**Review by divjotarora — COMMENTED** [review]
Couple of quick documentation comments, but the code changes LGTM.

File: mongo/database.go (hunk 6 above)
Comment (benjirewis) [reply mongo/database.go:184]: Moved to the end of the second paragraph.

File: mongo/options/serverapioptions.go (hunk 7 above)
Comment (benjirewis) [reply mongo/options/serverapioptions.go:20]: Sure, did it at the end of this comment. When a new API version is eventually added, we'll have to
- update this comment
- add a new ServerAPIVersion const
- update `Validate()` to handle the new version, and error with the correct message
- update `TestServerAPIVersion` in the driver

**Review by divjotarora — APPROVED** [review]
Fantastic work! LGTM

File: mongo/client_test.go [hunk 8]
```diff
@@ -361,18 +361,11 @@ func TestClient(t *testing.T) {
 			assert.Equal(t, convertedAPIOptions, client.serverAPI,
 				"mismatch in serverAPI; expected %v, got %v", convertedAPIOptions, client.serverAPI)
 		})
-		t.Run("failure with missing version", func(t *testing.T) {
```
Comment (kevinAlbs) [inline mongo/client_test.go]: It is nice that this is no longer an error scenario to consider!

**Review by kevinAlbs — APPROVED** [review]
Latest changes LGTM

**Pull request merged**